alibi_detect/utils/loading.py

```diff
@@ -0,0 +1,625 @@
+    ['kernel', 'kernel_b', 'sigma'],
+]
+
+# Fields to convert from str to np.dtype
+FIELDS_TO_DTYPE = [
+    ['preprocess_fn', 'dtype']
+]
+
+
+def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Detectors:
+    """
+    Load outlier, drift or adversarial detector.
+
+    Parameters
+    ----------
+    filepath
+        Load directory.
+
+    Returns
+    -------
+    Loaded outlier or adversarial detector object.
+    """
+    filepath = Path(filepath)
+    # If reference is a 'config.toml' itself, pass to new load function
+    if filepath.name == 'config.toml':
+        return _load_detector_config(filepath)
+
+    # Otherwise, if a directory, look for meta.dill, meta.pickle or config.toml inside it
+    elif filepath.is_dir():
+        files = [str(f.name) for f in filepath.iterdir() if f.is_file()]
+        if 'config.toml' in files:
+            return _load_detector_config(filepath.joinpath('config.toml'))
+        elif 'meta.dill' in files:
+            return load_detector_legacy(filepath, '.dill', **kwargs)
+        elif 'meta.pickle' in files:
+            return load_detector_legacy(filepath, '.pickle', **kwargs)
+        else:
+            raise ValueError('Neither meta.dill, meta.pickle or config.toml exist in {}.'.format(filepath))
+
+    # No other file types are accepted, so if not dir raise error
+    else:
+        raise ValueError("load_detector accepts only a filepath to a directory, or a config.toml file.")
+
+
+# TODO - will eventually become load_detector
+def _load_detector_config(filepath: Union[str, os.PathLike]) -> Detectors:
+    """
+    Loads a drift detector specified in a detector config dict. Validation is performed with pydantic.
+
+    Parameters
+    ----------
+    filepath
+        Directory containing the `config.toml` file.
+
+    Returns
+    -------
+    The instantiated detector.
+    """
+    # Load toml if needed
+    if isinstance(filepath, (str, os.PathLike)):
+        config_file = Path(deepcopy(filepath))
+        config_dir = config_file.parent
+        cfg = read_config(config_file)
+    else:
+        raise ValueError("`filepath` should point to a directory containing a 'config.toml' file.")
+
+    # Resolve and validate config
+    cfg = validate_config(cfg)
+    logger.info('Validated unresolved config.')
+    cfg = resolve_cfg(cfg, config_dir=config_dir)
+    cfg = validate_config(cfg, resolved=True)
+    logger.info('Validated resolved config.')
+
+    # Backend and detector type
+    detector_name = cfg.get('name')
+    backend = cfg.pop('backend')  # popping so that cfg left as kwargs + `name` when passed to _init_detector
+    if backend.lower() != 'tensorflow':
+        raise NotImplementedError('Loading detectors with PyTorch or sklearn backend is not yet supported.')
+
+    # Get x_ref
+    x_ref = cfg.pop('x_ref')
+    if isinstance(x_ref, str):  # If x_ref still a str, resolving must have failed
+        raise ValueError("Failed to resolve x_ref field.")
```

Comment:
Under what circumstances is this possible? Shouldn't the error be raised earlier by the resolution code?

Reply:
Removing this (and similar checks you noticed below), since we expect `resolve_config` to be passed a validated config. In other words, `validate_config(cfg, resolved=True)` will already have raised a `ValidationError` if `x_ref` is not an `np.ndarray`.